integrations/jta/jdbc/src/main/java/io/helidon/integrations/jta/jdbc/LocalXAResource.java

```diff
@@ -0,0 +1,1031 @@
+/*
+ * Copyright (c) 2022 Oracle and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.helidon.integrations.jta.jdbc;
+
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.util.EnumSet;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.function.UnaryOperator;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import javax.transaction.xa.XAException;
+import javax.transaction.xa.XAResource;
+import javax.transaction.xa.Xid;
+
+import io.helidon.integrations.jdbc.SQLRunnable;
+import io.helidon.integrations.jdbc.UncheckedSQLException;
+import io.helidon.integrations.jta.jdbc.ExceptionConverter.XARoutine;
+
+import static javax.transaction.xa.XAException.XAER_DUPID;
+import static javax.transaction.xa.XAException.XAER_INVAL;
+import static javax.transaction.xa.XAException.XAER_NOTA;
+import static javax.transaction.xa.XAException.XAER_PROTO;
+import static javax.transaction.xa.XAException.XAER_RMERR;
+import static javax.transaction.xa.XAException.XAER_RMFAIL;
+import static javax.transaction.xa.XAException.XA_RBROLLBACK;
+import static javax.transaction.xa.XAResource.TMENDRSCAN;
+import static javax.transaction.xa.XAResource.TMFAIL;
+import static javax.transaction.xa.XAResource.TMJOIN;
+import static javax.transaction.xa.XAResource.TMNOFLAGS;
+import static javax.transaction.xa.XAResource.TMONEPHASE;
+import static javax.transaction.xa.XAResource.TMRESUME;
+import static javax.transaction.xa.XAResource.TMSTARTRSCAN;
+import static javax.transaction.xa.XAResource.TMSUCCESS;
+import static javax.transaction.xa.XAResource.TMSUSPEND;
+import static javax.transaction.xa.XAResource.XA_OK;
+import static javax.transaction.xa.XAResource.XA_RDONLY;
+
+/**
+ * An {@link XAResource} that adapts an ordinary arbitrary {@link Connection} as much as possible to the {@link
+ * XAResource} contract.
+ *
+ * <p><strong>Note:</strong> instances of this class are lossless in the presence of one-phase commit operations and
+ * potentially lossy in the presence of two-phase commit operations.</p>
+ *
+ * <p>Instances of this class are safe for concurrent use by multiple threads.</p>
+ */
+final class LocalXAResource implements XAResource {
```

Comment:
Hello, reviewers. I'll annotate this PR myself with comments like this to help you navigate the XA world.